### PR TITLE
fix: remove unused components transitively

### PR DESCRIPTION
**Pull request by fb55**
## What/Why/How?

Removes components if their parent paths were removed. Loops until no more changes were made.

## Reference

Fixes https://github.com/Redocly/redocly-cli/issues/874

## Testing

Tests were updated to cover the new logic.

## Check yourself

- [ ] Code changed? - Tested with redoc/reference-docs/workflows (internal)
- [x] All new/updated code is covered with tests
- [ ] New package installed? - Tested in different environments (browser/node)

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

**Comment by fb55** [comment]
@tatomyr Please have a look!

**Comment by tatomyr** [comment]
@fb55 sure thing, will take a look!

**Review by tatomyr — CHANGES_REQUESTED** [review]
Sorry for the delay! 

Although I like the approach, there's one major issue with that solution to be fixed. 
Also, I left a few style change suggestions for the sake of readability (also applicable to the OAS2 variant). 

And one more thing. Whatever changes here should also apply to the corresponding rule [no-unused-components](https://github.com/Redocly/redocly-cli/blob/main/packages/core/src/rules/oas3/no-unused-components.ts). 
 
Again, thank you for your efforts. I hope you'll find my comments useful 🙂.

File: packages/core/src/decorators/oas3/remove-unused-components.ts [hunk 1]
```diff
@@ -49,17 +56,32 @@ export const RemoveUnusedComponents: Oas3Decorator = () => {
...
-              delete root.components[componentType];
+        const removedPaths: string[] = [];
+
+        let lastRemoveCount = 0;
+        do {
+          lastRemoveCount = 0;
+          for (const [path, { used, name, componentType }] of components) {
+            const isUsed = used?.some(
+              (location) =>
+                !removedPaths.some((removed) => location.absolutePointer.startsWith(removed))
```
[the remark this answers is not in the document]
Comment (fb55) [reply packages/core/src/decorators/oas3/remove-unused-components.ts:67]: Great catch! Adding a check for a trailing slash.

**Comment by fb55** [comment]
Thanks for the review, addressed all comments!

> And one more thing. Whatever changes here should also apply to the corresponding rule [no-unused-components](https://github.com/Redocly/redocly-cli/blob/main/packages/core/src/rules/oas3/no-unused-components.ts).

Do you mean introducing the additional data structures for tracking where a component is used?

**Comment by tatomyr** [comment]
> Do you mean introducing the additional data structures for tracking where a component is used?

Actually, at a second glance, I'm not sure we should apply the same logic in the rule itself... We're not removing transitive components but rather checking the current state. 

Let me have another look.

File: packages/core/src/decorators/oas2/remove-unused-components.ts [hunk 2]
```diff
@@ -16,15 +16,44 @@ export const RemoveUnusedComponents: Oas2Decorator = () => {
...
+  function removeUnusedComponents(root: Oas2Definition, removedPaths: string[]): number {
+    const removedLengthStart = removedPaths.length;
+
+    for (const [path, { usedIn, name, componentType }] of components) {
+      const used = usedIn?.some(
+        (location) =>
+          !removedPaths.some(
+            (removed) =>
+              location.absolutePointer.startsWith(removed) &&
+              (location.absolutePointer.length === removed.length ||
```
Comment (tatomyr) [inline packages/core/src/decorators/oas2/remove-unused-components.ts:34]: Isn't this the same as the following?
```
location.absolutePointer === removed ||
location.absolutePointer.startsWith(removed + '/') 

```

File: packages/core/src/decorators/oas2/remove-unused-components.ts [hunk 3]
```diff
@@ -16,15 +16,44 @@ export const RemoveUnusedComponents: Oas2Decorator = () => {
...
       componentType,
       name,
     });
   }
 
+  function removeUnusedComponents(root: Oas2Definition, removedPaths: string[]): number {
+    const removedLengthStart = removedPaths.length;
+
+    for (const [path, { usedIn, name, componentType }] of components) {
+      const used = usedIn?.some(
```
Comment (tatomyr) [inline packages/core/src/decorators/oas2/remove-unused-components.ts:29]:
```suggestion
      const used = usedIn.some(
```

Is the optional chaining still needed?

**Review by tatomyr — CHANGES_REQUESTED** [review]
WIth the latest changes it doesn't seem to remove transitive references anymore 😅.

File: packages/core/src/decorators/oas2/remove-unused-components.ts (hunk 2 above)
Comment (fb55) [reply packages/core/src/decorators/oas2/remove-unused-components.ts:34]: Yes, but it isn't comparing the strings twice & doesn't create a string copy. We are running through this loop a lot, so making it performant seems important.

File: packages/core/src/decorators/oas2/remove-unused-components.ts (hunk 3 above)
Comment (fb55) [reply packages/core/src/decorators/oas2/remove-unused-components.ts]: It is not!

**Comment by fb55** [comment]
> WIth the latest changes it doesn't seem to remove transitive references anymore 😅

Tests are passing, and the codebase I am looking at (~1000 auto-generated components with ~20% unused) is not showing any unused component errors after processing. But the reported number of removed components has gone down from ~220 to ~160 — not quite sure about what happened here.

**Comment by tatomyr** [comment]
> Tests are passing

Indeed they are. Sorry, mixed up with another errors 🙂.

File: packages/core/src/decorators/oas3/__tests__/remove-unused-components.test.ts [hunk 4]
```diff
@@ -168,4 +168,146 @@ describe('oas3 remove-unused-components', () => {
...
+        components:
+          schemas:
+            Unused: # <-- this will be removed correctly
+              type: integer
+            Transitive: # <-- this will be removed correctly
+              type: object
+              properties:
+                link:
+                  $ref: '#/components/schemas/Unused'
+            Used: # <-- this will be removed INCORRECTLY because the pointer starts with a previously removed path
```
Comment (tatomyr) [inline packages/core/src/decorators/oas3/__tests__/remove-unused-components.test.ts:258]: Could you change the comment? It *will not* be removed after the fix 🙂.

File: packages/core/src/decorators/oas2/__tests__/remove-unused-components.test.ts [hunk 5]
```diff
@@ -152,4 +152,77 @@ describe('oas2 remove-unused-components', () => {
...
+                '200':
+                  schema:
+                    $ref: '#/definitions/Used'
+              operationId: listPets
+              summary: List all pets
+        definitions:
+          Unused:
+            enum:
+              - 1
+              - 2
```
Comment (tatomyr) [inline packages/core/src/decorators/oas2/__tests__/remove-unused-components.test.ts:176]: Could you also clean up the tests from the needless fields like that one?

Comment (fb55) [reply packages/core/src/decorators/oas2/__tests__/remove-unused-components.test.ts:176]: This was copied from other test cases, and I am hesitant to include a larger cleanup here.

**Comment by fb55** [comment]
@tatomyr Anything I can help with to make this move forward?

**Comment by fb55** [comment]
Realised there was a comment I had not addressed, just did so!

File: packages/core/src/decorators/oas2/remove-unused-components.ts (hunk 2 above)
Comment (fb55) [reply packages/core/src/decorators/oas2/remove-unused-components.ts:36]: Added a comment!

**Review by lornajane — APPROVED** [review]
Changelog is good (approval from @Redocly/technical-writers)

**Review by tatomyr — APPROVED** [review]
Great job @fb55! 🔥 
Your changes will be available in the next release.

**Comment by fb55** [comment]
@tatomyr Thanks for shepherding this to the finish line!

**Pull request merged**